kobo-uncaged/kunc/kunc.go

```
@@ -0,0 +1,296 @@
+func (ku *koboUncaged) GetDeviceBookList() []uc.BookCountDetails {
+	bc := []uc.BookCountDetails{}
+	for _, md := range ku.k.MetadataMap {
+		lastMod := time.Now()
+		if md.LastModified != nil {
+			lastMod, _ = time.Parse(time.RFC3339, *md.LastModified)
+		}
+		bcd := uc.BookCountDetails{
+			UUID:         md.UUID,
+			Lpath:        md.Lpath,
+			LastModified: lastMod,
+		}
+		bcd.Extension = filepath.Ext(md.Lpath)
+		bc = append(bc, bcd)
+	}
+	//spew.Dump(bc)
+	return bc
+}
+
+// GetMetadataList sends complete metadata for the books listed in lpaths, or for
+// all books on device if lpaths is empty
+func (ku *koboUncaged) GetMetadataList(books []uc.BookID) []map[string]interface{} {
+	//spew.Dump(ku.k.MetadataMap)
+	//spew.Dump(books)
+	mdList := []map[string]interface{}{}
+	if len(books) > 0 {
+		for _, bk := range books {
+			cid := util.LpathToContentID(bk.Lpath, string(ku.k.ContentIDprefix))
+			fmt.Println(cid)
+			md := map[string]interface{}{}
+			//spew.Dump(ku.k.MetadataMap[cid])
+			mapstructure.Decode(ku.k.MetadataMap[cid], &md)
+			mdList = append(mdList, md)
+		}
+	} else {
+		for _, kmd := range ku.k.MetadataMap {
+			md := map[string]interface{}{}
+			//spew.Dump(kmd)
+			mapstructure.Decode(kmd, &md)
+			mdList = append(mdList, md)
+		}
+	}
+	return mdList
+}
+
+// GetDeviceInfo asks the client for information about the drive info to use
+func (ku *koboUncaged) GetDeviceInfo() uc.DeviceInfo {
+	return ku.k.DriveInfo
+}
+
+// SetDeviceInfo sets the new device info, as comes from calibre. Only the nested
+// struct DevInfo is modified.
+func (ku *koboUncaged) SetDeviceInfo(devInfo uc.DeviceInfo) {
+	ku.k.DriveInfo = devInfo
+	ku.k.SaveDeviceInfo()
+}
+
+// UpdateMetadata instructs the client to update their metadata according to the
+// new slice of metadata maps
+func (ku *koboUncaged) UpdateMetadata(mdList []map[string]interface{}) {
+	for _, md := range mdList {
+		koboMD := device.CreateKoboMetadata()
+		mapstructure.Decode(md, &koboMD)
```

```suggestion
		if err := mapstructure.Decode(md, &koboMD); err != nil {
			panic(err)
		}
```

The errors returned by decode all seem to be programming errors, and should probably not be silently ignored.

---

The metadata map list ultimately comes directly from calibre. Not sure if I want to be panicing here.

---

I'd rather not fail silently though.

---

It ultimately comes down to an interface issue again. I either log it, or do nothing at the moment.